raiden-cli/src/routes/channels.ts

```diff
@@ -52,7 +64,91 @@ async function openChannel(this: Cli, request: Request, response: Response, next
   } catch (error) {
     if (isInvalidParameterError(error)) {
       response.status(400).send(error.message);
-    } else if (error.code === 'INSUFFICIENT_FUNDS') {
+    } else if (isInvalidParameterError(error)) {
+      response.status(402).send(error.message);
+    } else if (isConflictError(error)) {
+      response.status(409).send(error.message);
+    } else {
+      next(error);
+    }
+  }
+}
+
+const allowedUpdateKeys = new Set(['state', 'total_deposit', 'total_withdraw']);
+function validateChannelUpdateBody(request: Request, response: Response, next: NextFunction) {
+  const intersec = Object.keys(request.body).filter((k) => allowedUpdateKeys.has(k));
+  if (intersec.length < 1)
+    return response
+      .status(400)
+      .send('one of "state" | "total_deposit" | "total_withdraw" operations required');
+  else if (intersec.length > 1)
+    return response
+      .status(409)
+      .send('more than one of "state" | "total_deposit" | "total_withdraw" requested');
+  if (
+    request.body.state &&
+    ![ApiChannelState.closed, ApiChannelState.settled].includes(request.body.state)
+  )
+    return response
+      .status(400)
+      .send('invalid "state" requested: must be one of "closed" | "settled"');
+  return next();
+}
+
+async function updateChannel(this: Cli, request: Request, response: Response, next: NextFunction) {
+  const token: string = request.params.tokenAddress;
+  const partner: string = request.params.partnerAddress;
+  try {
+    let channel = await this.raiden.channels$.pipe(first(), pluck(token, partner)).toPromise();
+    if (!channel) return response.status(404).send('channel not found');
+    if (request.body.state) {
+      if (request.body.state === ApiChannelState.closed) {
+        await this.raiden.closeChannel(token, partner);
+        channel = await this.raiden.channels$
+          .pipe(
+            pluck(token, partner),
+            first((channel) =>
+              [ChannelState.closed, ChannelState.settleable, ChannelState.settling].includes(
+                channel.state,
+              ),
+            ),
+          )
+          .toPromise();
+      } else if (request.body.state === ApiChannelState.settled) {
+        const promise = this.raiden.settleChannel(token, partner);
+        channel = await this.raiden.channels$.pipe(pluck(token, partner), first()).toPromise();
+        await promise;
+      }
+    } else if (request.body.total_deposit) {
+      await this.raiden.depositChannel(
+        token,
+        partner,
+        channel.ownDeposit.sub(request.body.total_deposit).mul(-1),
+      );
+      channel = await this.raiden.channels$
+        .pipe(
+          pluck(token, partner),
+          first((channel) => channel.ownDeposit.gte(request.body.total_deposit)),
```

Comment:
Again: why don't filter for the actual target deposit? Could it be that the target doesn't get reached by 100%? What happens if we chain two deposit requests? Couldn't it then be that the second gets the same response as the first one and never the actual update he requested for?
(same goes for the withdrawal below)

Reply:
Concurrent deposit is kind of not supported. These could cause some boring races which won't put the SDK in an unstable state, but will make one of the requests fail. This is expected, and this `gte` is just to ensure we either succeed (instead of getting stuck) if such race happens, or the user will get the rejected exception. There's no easy way to prevent the user calling it concurrently in the dApp, we would need to introduce locks or some kind of flags on the SDK, and I'd rather avoid that and leave it up to the user to know what they're doing, and possibly concurrent calls may succeed.